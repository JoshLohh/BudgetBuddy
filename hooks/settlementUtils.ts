// settlementUtils.ts
export function calculateBalances(members, expenses) {
  const paid = {};
  members.forEach(id => { paid[id] = 0; });
  expenses.forEach(exp => {
    paid[exp.paidBy] += parseFloat(exp.amount) || 0;
  });
  const total = expenses.reduce((sum, exp) => sum + (parseFloat(exp.amount) || 0), 0);
  const share = total / members.length;
  const balances = {};
  members.forEach(id => {
    balances[id] = +(paid[id] - share).toFixed(2);
  });
  return balances;
}

export function calculateSettlements(balances) {
  const creditors = [];
  const debtors = [];
  Object.entries(balances).forEach(([id, balance]) => {
    if (balance > 0.009) creditors.push({ id, amount: balance });
    else if (balance < -0.009) debtors.push({ id, amount: -balance });
  });
  creditors.sort((a, b) => b.amount - a.amount);
  debtors.sort((a, b) => b.amount - a.amount);

  const settlements = [];
  let i = 0, j = 0;
  while (i < debtors.length && j < creditors.length) {
    const d = debtors[i], c = creditors[j];
    const amount = Math.min(d.amount, c.amount);
    settlements.push({ from: d.id, to: c.id, amount: +amount.toFixed(2) });
    d.amount -= amount;
    c.amount -= amount;
    if (d.amount < 0.01) i++;
    if (c.amount < 0.01) j++;
  }
  return settlements;
}
