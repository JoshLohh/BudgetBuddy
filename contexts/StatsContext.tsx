import React, { createContext, useContext, useCallback, useEffect, useState } from 'react';
import { databases } from '@/lib/appwrite';
import { Query } from 'appwrite';
import { useUser } from '@/hooks/useUser';

type StatsContextType = {
  groupsCount: number;
  userExpensesCount: number;
  userTotalSpent: number;
  loading: boolean;
  refetchStats: () => Promise<void>;
};

const StatsContext = createContext<StatsContextType | undefined>(undefined);

export const StatsProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const { user } = useUser();
  const [groupsCount, setGroupsCount] = useState(0);
  const [userExpensesCount, setUserExpensesCount] = useState(0);
  const [userTotalSpent, setUserTotalSpent] = useState(0);
  const [loading, setLoading] = useState(false);

  const databaseId = process.env.EXPO_PUBLIC_APPWRITE_DATABASE_ID ?? '';
  const groupsCollectionId = process.env.EXPO_PUBLIC_APPWRITE_GROUPS_COLLECTION_ID ?? '';
  const expensesCollectionId = process.env.EXPO_PUBLIC_APPWRITE_EXPENSES_COLLECTION_ID ?? '';

  const fetchStats = useCallback(async () => {
    if (!user) {
      setGroupsCount(0);
      setUserExpensesCount(0);
      setUserTotalSpent(0);
      return;
    }
    setLoading(true);
    try {
      const groupsRes = await databases.listDocuments(
        databaseId,
        groupsCollectionId,
        [Query.contains('members', user.$id)]
      );
      setGroupsCount(groupsRes.documents.length);
      const groupIds = groupsRes.documents.map((doc: any) => doc.$id);

      if (groupIds.length === 0) {
        setUserExpensesCount(0);
        setUserTotalSpent(0);
        setLoading(false);
        return;
      }

      const expensesRes = await databases.listDocuments(
        databaseId,
        expensesCollectionId,
        [
          Query.equal('paidBy', user.$id),
          Query.equal('groupId', groupIds),
          Query.limit(100),
        ]
      );
      setUserExpensesCount(expensesRes.documents.length);
      const total = expensesRes.documents.reduce((sum: number, doc: any) => {
        const amt = parseFloat(String(doc.amount));
        return sum + (isNaN(amt) ? 0 : amt);
      }, 0);
      setUserTotalSpent(Number(total.toFixed(2)));
    } catch (error) {
      setGroupsCount(0);
      setUserExpensesCount(0);
      setUserTotalSpent(0);
      console.error('Failed to fetch stats:', error);
    } finally {
      setLoading(false);
    }
  }, [user, databaseId, groupsCollectionId, expensesCollectionId]);

  useEffect(() => {
    fetchStats();
  }, [fetchStats]);

  return (
    <StatsContext.Provider
      value={{
        groupsCount,
        userExpensesCount,
        userTotalSpent,
        loading,
        refetchStats: fetchStats,
      }}
    >
      {children}
    </StatsContext.Provider>
  );
};

export const useStats = () => {
  const ctx = useContext(StatsContext);
  if (!ctx) throw new Error('useStats must be used within StatsProvider');
  return ctx;
};
