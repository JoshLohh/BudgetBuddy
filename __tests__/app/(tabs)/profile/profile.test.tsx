/**
 * @file Profile.test.tsx
 * Test suite for Profile.tsx
 *
 * Covers key functionalities including:
 * - Rendering user profile information
 * - Editing profile data and saving updates
 * - Handling avatar image upload with permission checks and errors
 * - Logout functionality and navigation
 * - Loading states and timeout reload behavior
 *
 * Uses mocks for:
 * - useUser hook (profile data, update, logout)
 * - useStats hook (group/expense stats)
 * - Expo Router (navigation mocks)
 * - Expo ImagePicker (permissions and image launch)
 * - Global fetch (for avatar upload)
 */

import React from 'react';
import {
  render,
  fireEvent,
  waitFor,
  act,
} from '@testing-library/react-native';
import Profile from '@/app/(tabs)/profile/profile';
import { useUser } from '@/hooks/useUser';
import { useStats } from '@/contexts/StatsContext';
import * as ImagePicker from 'expo-image-picker';
import { useRouter } from 'expo-router';

jest.mock('@/hooks/useUser');
jest.mock('@/contexts/StatsContext');
jest.mock('expo-image-picker');
jest.mock('expo-router');

global.alert = jest.fn(); // Mock global alert to suppress dialogs in tests

describe('Profile Component', () => {
  const mockUser = { email: 'test@example.com', $id: 'user1' };
  const mockProfile = {
    username: 'TestUser',
    email: 'test@example.com',
    bio: 'Hello world',
    avatar: 'https://example.com/avatar.jpg',
  };
  const mockStats = {
    groupsCount: 3,
    userExpensesCount: 10,
    userTotalSpent: 150.75,
    loading: false,
    refetchStats: jest.fn(),
  };
  const mockRefetchProfile = jest.fn().mockResolvedValue(mockProfile);
  const mockUpdateProfile = jest.fn().mockResolvedValue(undefined);
  const mockLogout = jest.fn().mockResolvedValue(undefined);
  const mockRouterReplace = jest.fn();
  const mockRouterPush = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();

    // Mock useUser hook
    (useUser as jest.Mock).mockReturnValue({
      user: mockUser,
      profile: mockProfile,
      updateProfile: mockUpdateProfile,
      logout: mockLogout,
      authChecked: true,
      refetchProfile: mockRefetchProfile,
    });

    // Mock useStats hook
    (useStats as jest.Mock).mockReturnValue(mockStats);

    // Mock expo-router hooks
    (useRouter as jest.Mock).mockReturnValue({
      replace: mockRouterReplace,
      push: mockRouterPush,
    });

    // Mock Expo ImagePicker permissions granted and launch results
    (ImagePicker.requestMediaLibraryPermissionsAsync as jest.Mock).mockResolvedValue({ granted: true });
    (ImagePicker.launchImageLibraryAsync as jest.Mock).mockResolvedValue({
      canceled: false,
      assets: [{ uri: 'file:///test-image.jpg', mimeType: 'image/jpeg' }],
    });

    // Mock fetch for file upload with successful JSON response
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.resolve({ $id: 'file123' }),
      } as Response)
    ) as unknown as typeof fetch;
  });

  it('renders profile information correctly', () => {
    const { getByText } = render(<Profile />);
    expect(getByText('TestUser')).toBeTruthy();
    expect(getByText('Hello world')).toBeTruthy();

    // Separate assertions for split Text nodes in stats section
    expect(getByText('3')).toBeTruthy();
    expect(getByText('Groups')).toBeTruthy();

    expect(getByText('$150.75')).toBeTruthy();
    //expect(getByText('150.75')).toBeTruthy();
    expect(getByText('Total Spent')).toBeTruthy();

    expect(getByText('10')).toBeTruthy();
    expect(getByText('# Expenses')).toBeTruthy();
  });

  it('toggles editing mode and populates inputs', () => {
    const { getByText, getByDisplayValue } = render(<Profile />);
    fireEvent.press(getByText('Edit Profile'));
    expect(getByDisplayValue('TestUser')).toBeTruthy();
    expect(getByText('test@example.com')).toBeTruthy();
    expect(getByDisplayValue('Hello world')).toBeTruthy();
  });

  it('saves profile changes and exits edit mode on success', async () => {
    const { getByText } = render(<Profile />);
    fireEvent.press(getByText('Edit Profile'));
    fireEvent.press(getByText('Save Changes'));
    await waitFor(() => {
      expect(mockUpdateProfile).toHaveBeenCalledWith(expect.objectContaining({
        username: 'TestUser',
        email: 'test@example.com',
        bio: 'Hello world',
      }));
      expect(getByText('Edit Profile')).toBeTruthy(); // Checks edit button shown after save
    });
  });

  it('shows error message if save profile fails', async () => {
    mockUpdateProfile.mockRejectedValueOnce(new Error('Save failed'));
    const { getByText } = render(<Profile />);
    fireEvent.press(getByText('Edit Profile'));
    fireEvent.press(getByText('Save Changes'));
    await waitFor(() => {
      expect(getByText('Save failed')).toBeTruthy();
    });
  });

  it('handles avatar upload permission denied', async () => {
    (ImagePicker.requestMediaLibraryPermissionsAsync as jest.Mock).mockResolvedValueOnce({ granted: false });
    global.alert = jest.fn();
    const { getByLabelText } = render(<Profile />);
    // Make sure there is an accessibilityLabel for avatar change button in your Profile component!
    const avatarButton = getByLabelText('Change Avatar');
    await act(async () => {
      fireEvent.press(avatarButton);
    });
    expect(global.alert).toHaveBeenCalledWith('Permission to access gallery is required!');
  });

  it('uploads avatar image successfully and updates profile', async () => {
    const { getByLabelText } = render(<Profile />);
    const avatarButton = getByLabelText('Change Avatar');
    await act(async () => {
      fireEvent.press(avatarButton);
    });
    expect(global.fetch).toHaveBeenCalled();
    expect(mockUpdateProfile).toHaveBeenCalledWith(expect.objectContaining({
      avatar: expect.stringContaining('storage/buckets'),
    }));
    expect(mockRefetchProfile).toHaveBeenCalled();
  });

  it('alerts on avatar upload server error', async () => {
    // Mock fetch to trigger catch for invalid JSON
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: true,
        json: () => Promise.reject('server error'),
      } as Response)
    ) as unknown as typeof fetch;

    global.alert = jest.fn();
    const { getByLabelText } = render(<Profile />);
    const avatarButton = getByLabelText('Change Avatar');

    await act(async () => {
      fireEvent.press(avatarButton);
    });

    expect(global.alert).toHaveBeenCalledWith('Upload failed: invalid response from server.');
  });

  it('alerts on avatar upload non-ok response', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        ok: false,
        json: () => Promise.resolve({ message: 'Upload error' }),
      } as Response)
    ) as unknown as typeof fetch;

    global.alert = jest.fn();
    const { getByLabelText } = render(<Profile />);
    const avatarButton = getByLabelText('Change Avatar');

    await act(async () => {
      fireEvent.press(avatarButton);
    });

    expect(global.alert).toHaveBeenCalledWith('Upload failed: Upload error');
  });

  it('calls logout and navigates to root on logout button press', async () => {
    const { getByText } = render(<Profile />);
    fireEvent.press(getByText('Logout'));
    await waitFor(() => {
      expect(mockLogout).toHaveBeenCalled();
      expect(mockRouterReplace).toHaveBeenCalledWith('/');
    });
  });

  it('shows loading state with reload button on profile load timeout', async () => {
    (useUser as jest.Mock).mockReturnValue({
      user: null,
      profile: null,
      updateProfile: mockUpdateProfile,
      logout: mockLogout,
      authChecked: true,
      refetchProfile: mockRefetchProfile,
    });

    jest.useFakeTimers();
    const { getByText } = render(<Profile />);

    expect(getByText(/Loading profile/)).toBeTruthy();

    act(() => {
      jest.advanceTimersByTime(30000);
    });

    expect(getByText('Still loading? Tap reload below.')).toBeTruthy();

    const reloadButton = getByText('Reload');
    fireEvent.press(reloadButton);

    await waitFor(() => {
      expect(mockRefetchProfile).toHaveBeenCalled();
      expect(mockStats.refetchStats).toHaveBeenCalled();
    });

    jest.useRealTimers();
  });

  it('matches snapshot (non-edit mode)', () => {
    const tree = render(<Profile />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot (edit mode)', () => {
    const { getByText } = render(<Profile />);
    fireEvent.press(getByText('Edit Profile'));
    const tree = render(<Profile />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
