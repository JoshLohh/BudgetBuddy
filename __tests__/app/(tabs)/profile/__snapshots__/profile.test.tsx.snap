// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component matches snapshot (edit mode) 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "flex": 1,
        "justifyContent": "flex-start",
        "paddingBottom": 16,
        "paddingHorizontal": 20,
        "paddingTop": 36,
      },
    ]
  }
  testID="themed-view"
>
  <View
    style={
      {
        "height": 40,
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "marginBottom": 8,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#0a7ea4",
              "fontSize": 28,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        î¹¬
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 18,
      }
    }
  >
    <View
      accessibilityLabel="Change Avatar"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <ViewManagerAdapter_ExpoImage
        borderColor={4294967295}
        borderRadius={75}
        borderWidth={2}
        containerViewRef={"[React.ref]"}
        contentFit="cover"
        contentPosition={
          {
            "left": "50%",
            "top": "50%",
          }
        }
        height={150}
        marginRight={18}
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={[]}
        source={
          [
            {
              "uri": "https://example.com/avatar.jpg",
            },
          ]
        }
        style={
          {
            "borderColor": "#fff",
            "borderRadius": 75,
            "borderWidth": 2,
            "height": 150,
            "marginRight": 18,
            "width": 150,
          }
        }
        transition={null}
        width={150}
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            undefined,
            {
              "fontSize": 28,
              "fontWeight": "bold",
              "lineHeight": 32,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "marginBottom": 0,
            },
          ]
        }
      >
        TestUser
      </Text>
    </View>
  </View>
  <View>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "fontSize": 15,
            "marginBottom": 6,
            "marginLeft": 12,
          },
        ]
      }
    >
      Hello world
    </Text>
    <View
      style={
        {
          "height": 8,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      {
        "borderRadius": 12,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 24,
        "padding": 18,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        3
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        Groups
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        $
        150.75
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        Total Spent
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        10
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        # Expenses
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderRadius": 6,
            "marginVertical": 10,
            "padding": 18,
          },
          {
            "backgroundColor": "#0a7ea4",
          },
          false,
          {
            "borderRadius": 8,
            "marginHorizontal": 16,
            "paddingVertical": 12,
            "width": "50%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#f2f2f2",
              "textAlign": "center",
            },
          ]
        }
      >
        Edit Profile
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "flex-end",
        "paddingBottom": 24,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderRadius": 6,
            "marginVertical": 10,
            "padding": 18,
          },
          {
            "backgroundColor": "#0a7ea4",
          },
          false,
          {
            "backgroundColor": "#cb4d31",
            "borderRadius": 8,
            "paddingVertical": 12,
            "width": "50%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#f2f2f2",
              "textAlign": "center",
            },
          ]
        }
      >
        Logout
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "height": 75,
        "width": "100%",
      }
    }
  />
</View>
`;

exports[`Profile Component matches snapshot (non-edit mode) 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    [
      {
        "backgroundColor": "#fff",
      },
      {
        "flex": 1,
        "justifyContent": "flex-start",
        "paddingBottom": 16,
        "paddingHorizontal": 20,
        "paddingTop": 36,
      },
    ]
  }
  testID="themed-view"
>
  <View
    style={
      {
        "height": 40,
        "width": "100%",
      }
    }
  />
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "flex-end",
        "marginBottom": 8,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          [
            {
              "color": "#0a7ea4",
              "fontSize": 28,
            },
            undefined,
            {
              "fontFamily": "ionicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            {},
          ]
        }
      >
        î¹¬
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "marginBottom": 18,
      }
    }
  >
    <View
      accessibilityLabel="Change Avatar"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <ViewManagerAdapter_ExpoImage
        borderColor={4294967295}
        borderRadius={75}
        borderWidth={2}
        containerViewRef={"[React.ref]"}
        contentFit="cover"
        contentPosition={
          {
            "left": "50%",
            "top": "50%",
          }
        }
        height={150}
        marginRight={18}
        nativeViewRef={"[React.ref]"}
        onError={[Function]}
        onLoad={[Function]}
        onLoadStart={[Function]}
        onProgress={[Function]}
        placeholder={[]}
        source={
          [
            {
              "uri": "https://example.com/avatar.jpg",
            },
          ]
        }
        style={
          {
            "borderColor": "#fff",
            "borderRadius": 75,
            "borderWidth": 2,
            "height": 150,
            "marginRight": 18,
            "width": 150,
          }
        }
        transition={null}
        width={150}
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            undefined,
            {
              "fontSize": 28,
              "fontWeight": "bold",
              "lineHeight": 32,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "marginBottom": 0,
            },
          ]
        }
      >
        TestUser
      </Text>
    </View>
  </View>
  <View>
    <Text
      style={
        [
          {
            "color": "#11181C",
          },
          {
            "fontSize": 16,
            "lineHeight": 24,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
          {
            "fontSize": 15,
            "marginBottom": 6,
            "marginLeft": 12,
          },
        ]
      }
    >
      Hello world
    </Text>
    <View
      style={
        {
          "height": 8,
          "width": "100%",
        }
      }
    />
  </View>
  <View
    style={
      {
        "borderRadius": 12,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginBottom": 24,
        "padding": 18,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        3
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        Groups
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        $
        150.75
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        Total Spent
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flex": 1,
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 17,
              "fontWeight": "bold",
            },
          ]
        }
      >
        10
      </Text>
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 12,
              "marginTop": 2,
            },
          ]
        }
      >
        # Expenses
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "alignItems": "center",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderRadius": 6,
            "marginVertical": 10,
            "padding": 18,
          },
          {
            "backgroundColor": "#0a7ea4",
          },
          false,
          {
            "borderRadius": 8,
            "marginHorizontal": 16,
            "paddingVertical": 12,
            "width": "50%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#f2f2f2",
              "textAlign": "center",
            },
          ]
        }
      >
        Edit Profile
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "flex": 1,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "flex-end",
        "paddingBottom": 24,
        "width": "100%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "borderRadius": 6,
            "marginVertical": 10,
            "padding": 18,
          },
          {
            "backgroundColor": "#0a7ea4",
          },
          false,
          {
            "backgroundColor": "#cb4d31",
            "borderRadius": 8,
            "paddingVertical": 12,
            "width": "50%",
          },
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#11181C",
            },
            {
              "fontSize": 16,
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            {
              "color": "#f2f2f2",
              "textAlign": "center",
            },
          ]
        }
      >
        Logout
      </Text>
    </View>
  </View>
  <View
    style={
      {
        "height": 75,
        "width": "100%",
      }
    }
  />
</View>
`;
